onnxscript/function_libs/torch_lib/ops/core.py

```diff
@@ -3164,36 +3164,41 @@ def aten_isclose(
 def aten_isfinite(self: TFloatOrBFloat16) -> BOOL:
     """isfinite(Tensor self) -> Tensor"""
 
-    not_inf = op.Not(op.IsInf(self))
+    self = op.Cast(self, to=FLOAT.dtype)  # Make this function support all real types
```

Comment:
This op.Cast makes no sense to me. op.IsInf simply doesn't support float16. I would change input annotation to TFloatUnlessFloat16

Reply:
Yes, your consideration is also reasonable. Although I think I'm right, while I cannot say you are wrong. :)
If the annotation TFloaUnlessFloat16 can help when programmatically convert the model to fp16, I'm OK.

Reply:
Although OnnxFunction doesn't have to have the same spec as ONNX, our annotation rule rn is to annotate them by the intersection of type constraints of used Ops. Adding Cast here to expand the function support is really weird. Even if we wanted to expand the dtype support  by upcast/downcast, we should do it in converter side IMO.

Reply:
@justinchuby and I discussed about adding upcast/downcast into the dispatcher to support more dtype that are out of ONNX spec. Maybe it's similar to what you wanted to do here?

Reply:
I wasn't in the discussion, could you elaborate more detail about why we definitely need to support fp16 on a function using non-support fp16 op?

Reply:
Similar thought but different way. I think FP16 is special, because almost all the users will convert fp32 model to fp16 (because I maintain the convert_fp32_to_fp16 tool in onnxmltools, so I know). For other dtypes such as int64 or double, I think the priority is low, or even don't need to care about.

Reply:
I see. Could you find somewhere good to put a NOTE something like "FP16 is special case that we support with inserting op.Cast inside function."?